### PR TITLE
Chime4

### PR DESCRIPTION
**Pull request by JorisCos**
### About this PR
This PR introduces a recipe for evaluation purposes only, on real noisy data. 
It uses the data from the CHiME 4 challenge (an adaptation from the CHiME 3 challenge).

### Design
I am unsure about the design of the dataset and the recipe : 

- The dataset mimics the behaviour of other datasets by returning mixture and sources(here fake sources that are actually a copy of the mixture) 
- The recipe only contains scripts for the evaluation it might be worth to make it more general by avoiding to limit its application to only one model

### CODE REVIEW

**Comment by mpariente** [comment]
I'm undecided about this. 
On the one hand, having data prep and eval script for Chime4 is cool, and it's an example for other models but we could probably make more modular than this.  

But to make it more modular, we should attach the data prep to the dataloader (needs refactoring of the data part of Asteroid, which we want to do), and we should enable evaluating any model on any dataset (also needs refactoring). 

I'd vote in favour of merging it, because it diversifies the examples that we have (both for the users and things in mind when we refactor). 

@jonashaag @popcornell

**Comment by jonashaag** [comment]
Yes, I'd say merge it, as it's not introducing any new way of doing things, so it doesn't add much work to our refactoring plate.

Did you forget to rename the "chime3" variables and CLI flag or is that on purpose?

**Comment by JorisCos** [comment]
> Yes, I'd say merge it, as it's not introducing any new way of doing things, so it doesn't add much work to our refactoring plate.
> 
> Did you forget to rename the "chime3" variables and CLI flag or is that on purpose?

That's on purpose because CHiME 4 uses data from CHiME 3 when you download the data you actually get a directory named CHiME3. But maybe it's confusing.

**Comment by mpariente** [comment]
I guess it will be clear once people have the data from CHiME3 downloaded.

File: asteroid/data/chime4_dataset.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import pandas as pd
+import soundfile as sf
+import torch
+from torch.utils.data import Dataset, DataLoader
+import random as random
+import os
+
+
+class CHiME4(Dataset):
```
Comment (mpariente) [inline asteroid/data/chime4_dataset.py:9]: Let's call this `CHiME4Dataset`.

File: asteroid/data/chime4_dataset.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        # Read the mixture
+        mixture, _ = sf.read(self.mixture_path, dtype="float32", start=start, stop=stop)
+        # Convert to torch tensor
+        mixture = torch.from_numpy(mixture)
+        mock_source = torch.vstack([mixture])
+        if self.return_id:
+            id1 = row.ID
+            return mixture, mock_source, [id1]
+        return mixture, mock_source
```
Comment (mpariente) [inline asteroid/data/chime4_dataset.py:72]: Let's not return the mock_source.

File: egs/chime4/ConvTasNet/eval.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    torch.no_grad().__enter__()
+    for idx in tqdm(range(len(test_set))):
+        # Forward the network on the mixture.
+        mix, sources, ids = test_set[idx]
+        mix, sources = tensors_to_device([mix, sources], device=model_device)
+        est_sources = model(mix.unsqueeze(0))
+        mix_np = mix.cpu().data.numpy()
+        sources_np = sources.cpu().data.numpy()
+        est_sources_np = est_sources.squeeze(0).cpu().data.numpy()
+        est_sources_np *= np.max(np.abs(mix_np)) / np.max(np.abs(est_sources_np))
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/eval.py:104]: Remove the code related to `sources`.

File: egs/chime4/ConvTasNet/eval.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        # For each utterance, we get a dictionary with the mixture path,
+        # the input and output metrics
+        utt_metrics = {"mix_path": test_set.mixture_path}
+        utt_metrics.update(
+            **wer_tracker(
+                mix=mix_np,
+                clean=sources_np,
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/eval.py:111]: Extend `WerTracker` to work without clean reference. And remove clean.

File: egs/chime4/ConvTasNet/eval.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        ldf = all_metrics_df[metric_name] - all_metrics_df[input_metric_name]
+        final_results[metric_name] = all_metrics_df[metric_name].mean()
+        final_results[metric_name + "_imp"] = ldf.mean()
+
+    print("Overall metrics :")
+    pprint(final_results)
+    if conf["compute_wer"]:
+        print("\nWER report")
+        wer_card = wer_tracker.final_report_as_markdown()
+        print(wer_card)
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/eval.py:155]: Only print things from `mixture` and `estimates`. 
And potential improvement.

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        write_dataframe(df, df_2, subset, origin)
+
+
+def create_dataframe(chime3_dir, c3_annot_file, c4_annot_file, subset, origin):
+    # Empty list for DataFrame creation
+    row_list = []
+    row_list_2 = []
+    for row in c3_annot_file.itertuples():
+        speaker = row.speaker
+        ID = row.wsj_name
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/local/create_metadata.py:53]: Which type of ID is this? Utterance ID? 
Can we change the name to clarify, and put everything lower case?

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        env = row.environment
+        # if we are not dealing with tr subset
+        if "tr" not in subset:
+            mixture_path = c4_annot_file[c4_annot_file["path"].str.contains(ID + "_" + env)].values[
+                0
+            ][0]
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/local/create_metadata.py:59]: I know this is black, but this is unreadable. 
We can call `c4_anno` and `c3_anno` instead of `c4_annot_file`, or smth else, as you want.

File: egs/chime4/ConvTasNet/run.sh [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+eval_use_gpu=0
+# Need to --compute_wer 1 --eval_mode max to be sure the user knows all the metrics
+# are for the all mode.
+compute_wer=1
+
+# Choice for the ASR model whether trained on clean or noisy data. One of clean or noisy
+asr_type=noisy
+
+test_dir=data/test
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/run.sh:34]: Does this need to be overwritten? 
If it doesn't, then add it after utils/parse_options.sh

File: egs/chime4/ConvTasNet/run.sh [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    fi
+  fi
+
+  $python_path eval.py \
+    --exp_dir $exp_dir \
+    --test_dir $test_dir \
+  	--use_gpu $eval_use_gpu \
+  	--compute_wer $compute_wer \
+  	--asr_type $asr_type
+fi
```
Comment (mpariente) [inline egs/chime4/ConvTasNet/run.sh:65]: Fix indents, but otherwise OK

File: egs/chime4/README.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+It was released in 2016 and revisits the datasets originally recorded for CHiME-3.
+
+All data and information are available [here](http://spandh.dcs.shef.ac.uk/chime_challenge/CHiME4/index.html).
+
+For now, this recipe only deals with the `real_1_ch_track` part of the dataset. 
+As the channel to use for the training set wasn't defined by 
+the challenge's rules, we will set it randomly.
+
+NOTE :
```
Comment (mpariente) [inline egs/chime4/README.md:13]:
```suggestion
**Note :**
```

File: egs/chime4/README.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+All data and information are available [here](http://spandh.dcs.shef.ac.uk/chime_challenge/CHiME4/index.html).
+
+For now, this recipe only deals with the `real_1_ch_track` part of the dataset. 
+As the channel to use for the training set wasn't defined by 
+the challenge's rules, we will set it randomly.
+
+NOTE : 
+This dataset uses real noisy data. This means the clean speech from the noisy
+utterances is not available. This makes it not suitable for the usual training
```
Comment (mpariente) [inline egs/chime4/README.md:15]:
```suggestion
utterances is not available. This makes it unsuitable for the usual training 
```

File: egs/chime4/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+The CHiME-4 dataset is part of the 4th CHiME speech separation and recognition challenge.
+
+It was released in 2016 and revisits the datasets originally recorded for CHiME-3.
+
+All data and information are available [here](http://spandh.dcs.shef.ac.uk/chime_challenge/CHiME4/index.html).
+
+For now, this recipe only deals with the `real_1_ch_track` part of the dataset. 
+As the channel to use for the training set wasn't defined by 
+the challenge's rules, we will set it randomly.
```
Comment (mpariente) [inline egs/chime4/README.md:11]: I don't understand that, can you explain please?

**Review by mpariente — COMMENTED** [review]
Mainly good to do but needs a last polishing. 

I feel the create_metadata.py file is not so readable. Mainly because of the function flow. I'd have to think about what I'd reproach to the design exactly.

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    # Get CHiME-3 annotation files
+    c3_annot_files = [
+        f for f in glob(os.path.join(chime3_dir, "data", "annotations", "*real*.json"))
+    ]
+    # Get CHiME-4 annotation files
+    c4_annot_files = [
+        f for f in glob(os.path.join(chime3_dir, "data", "annotations", "*real*.list"))
+    ]
+    for c3_annot_file_path in c3_annot_files:
+        # Read CHiME-3 annotation file
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:31]: I think it could be helpful to move the entire `for` loop body into a separate function so as to make it obvious that all loop iterations are entirely independent.

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        # subsets : "tr" "dt" "et" origin "real" or "simu"
+        subset, origin = os.path.split(c3_annot_file_path)[1].replace(".json", "").split("_")
+        # Look for associated CHiME-4 file
+        if c3_annot_file_path.replace(".json", "_1ch_track.list") in c4_annot_files:
+            # Read CHiME-4 annotation file
+            c4_annot_file = pd.read_csv(
+                c3_annot_file_path.replace(".json", "_1ch_track.list"), header=None, names=["path"]
+            )
+        else:
+            c4_annot_file = None
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:42]: This case isn't handled in the `create_dataframe` function

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    ]
+    # Get CHiME-4 annotation files
+    c4_annot_files = [
+        f for f in glob(os.path.join(chime3_dir, "data", "annotations", "*real*.list"))
+    ]
+    for c3_annot_file_path in c3_annot_files:
+        # Read CHiME-3 annotation file
+        c3_annot_file = pd.read_json(c3_annot_file_path)
+        # subsets : "tr" "dt" "et" origin "real" or "simu"
+        subset, origin = os.path.split(c3_annot_file_path)[1].replace(".json", "").split("_")
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:34]: I think the entire code could benefit from using the `pathlib` API. Eg here, `c3_annot_file_path.with_suffix("").name.split("_")`

Also I think it could be helpful to add a comment with a file name example everywhere filenames are parsed or constructed, for example

```
# Extract subset and origin from /foo/bar/<subset>_<origin>.json
```

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        # Look for associated CHiME-4 file
+        if c3_annot_file_path.replace(".json", "_1ch_track.list") in c4_annot_files:
+            # Read CHiME-4 annotation file
+            c4_annot_file = pd.read_csv(
+                c3_annot_file_path.replace(".json", "_1ch_track.list"), header=None, names=["path"]
+            )
+        else:
+            c4_annot_file = None
+        df, df_2 = create_dataframe(chime3_dir, c3_annot_file, c4_annot_file, subset, origin)
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:43]: Could there be better names than `df` and `df_2`?

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        write_dataframe(df, df_2, subset, origin)
+
+
+def create_dataframe(chime3_dir, c3_annot_file, c4_annot_file, subset, origin):
+    # Empty list for DataFrame creation
+    row_list = []
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:49]: Again, is there a more helpful name than "row" and "row 2" here?

File: egs/chime4/ConvTasNet/local/create_metadata.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    row_list_2 = []
+    for row in c3_annot_file.itertuples():
+        speaker = row.speaker
+        ID = row.wsj_name
+        env = row.environment
+        # if we are not dealing with tr subset
+        if "tr" not in subset:
+            mixture_path = c4_annot_file[c4_annot_file["path"].str.contains(ID + "_" + env)].values[
```
Comment (jonashaag) [inline egs/chime4/ConvTasNet/local/create_metadata.py:57]: Here, and also in the `else` case below, it could be helpful to add a comment with an example of what we're looking for in the annotation data frame, like the I suggested with the path splitting above

**Review by jonashaag — COMMENTED** [review]
Manuel asked for some review of `create_metadata.py` so here we go :)

File: egs/chime4/README.md [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+NOTE : 
+This dataset uses real noisy data. This means the clean speech from the noisy
+utterances is not available. This makes it not suitable for the usual training 
+procedure.
+
+
+
+**References**
+~~~BibTeX
```
[the remark this answers is not in the document]
Comment (JorisCos) [reply egs/chime4/README.md:21]: It produces the same result I believe

File: egs/chime4/README.md (hunk 12 above)
Comment (JorisCos) [reply egs/chime4/README.md:11]: When the CHiME4 challenge was released they gave the mics to use for the test set and the dev set for the `1-channel track ` but they didn't give instructions for the training set. It's up to the users to chose their mics among the 6 that are provided in the original CHiME3 dataset.

Comment (mpariente) [reply egs/chime4/README.md:11]: For training, this seems fine, but not for evaluation..
IIRC, channel 5 was used? I'm not sure, maybe we can have a look in ESPNet. 
@sw005320 do you know which channel to use for evaluation in Chime4 please?

Comment (JorisCos) [reply egs/chime4/README.md:11]: For evaluation we use the channel provided by the challenge. Sorry if I was unclear, we have the channel to use for the evaluation set and the dev set but not for the training set
